File: android/hakka-network-noop/src/main/kotlin/com/noodleapps/hakka/HakkaInterceptor.kt
package com.noodleapps.hakka

import okhttp3.Interceptor
import okhttp3.Response

/**
 * No-op OkHttp [Interceptor] that passes all requests through without capturing.
 * Same public API as [com.noodleapps.hakka.HakkaInterceptor] — swap by dependency name.
 */
class HakkaInterceptor private constructor(
    private val config: HakkaConfig,
    listeners: List<HakkaListener>,
    sinks: List<RecordSink>,
) : Interceptor, AutoCloseable {

    val logStore = LogStore(config)

    /**
     * No-op plugin registry — [PluginRegistry.use] is idempotent and calls plugin setup
     * with a no-op context. API parity with `hakka-network`.
     */
    val plugins: PluginRegistry = PluginRegistry()

    /**
     * No-op config update. Accepts the same signature as the real interceptor.
     * Mirrors `updateConfig(_:)` on iOS.
     */
    fun updateConfig(transform: (HakkaConfig) -> HakkaConfig) = Unit

    /**
     * Returns the retention policy derived from the initial config.
     * Mirrors [retentionPolicy] on iOS.
     */
    val retentionPolicy: RetentionPolicy get() = RetentionPolicy.from(config)

    /** Returns null — no timing capture in noop. */
    fun eventListenerFactory(): HakkaEventListener.Factory? = null

    /** Returns empty list — no in-flight tracking in noop. */
    fun inFlightRequests(): List<NetworkRequest> = emptyList()

    /** Returns empty list — no request capture in noop. */
    fun recentRequests(count: Int = 20): List<NetworkRequest> = emptyList()

    /** Returns true immediately — no queued capture work in noop. */
    fun flushCaptureProcessing(timeoutMs: Long = 5_000L): Boolean = true

    /** Returns true immediately — no sink work in noop. */
    fun flushSinks(timeoutMs: Long = 5_000L): Boolean = true

    /** Always zero in noop. */
    fun droppedSinkRecords(): Long = 0

    /** Builds an inert health report with no captured records. */
    fun healthReport(
        timestampMs: Long = System.currentTimeMillis(),
        sessionId: String? = null,
        tags: Map<String, String> = emptyMap(),
    ): HealthReportRecord =
        HealthReportGenerator.fromRecords(
            records = emptyList(),
            options = HealthReportBuildOptions(
                timestampMs = timestampMs,
                sessionId = sessionId,
                tags = tags + mapOf(
                    "component.capture.inFlightCount" to "0",
                    "component.storage.count" to "0",
                    "component.storage.maxCount" to config.maxRequests.toString(),
                    "component.redaction.headerCount" to config.redactHeaders.size.toString(),
                    "component.redaction.queryItemCount" to config.sensitiveQueryItems.size.toString(),
                    "component.redaction.bodyFieldCount" to config.sensitiveBodyFields.size.toString(),
                ),
                componentHealth = mapOf(
                    "capture" to ComponentHealth(status = "disabled"),
                    "storage" to ComponentHealth(status = "disabled"),
                    "redaction" to ComponentHealth(
                        status = if (
                            config.redactHeaders.isNotEmpty() ||
                            config.sensitiveQueryItems.isNotEmpty() ||
                            config.sensitiveBodyFields.isNotEmpty()
                        ) "configured" else "disabled",
                    ),
                    "sink" to ComponentHealth(status = "disabled", droppedCount = 0),
                ),
            ),
        )

    /** Always returns a zeroed summary — no requests captured in noop. */
    fun networkMetricsSummary(): NetworkMetricsSummary = logStore.metricsSummary()

    /** Returns true immediately — no capture processor in noop. */
    fun shutdownCaptureProcessing(timeoutMs: Long = 1_000L): Boolean = true

    /** Returns true immediately — no sink executor in noop. */
    fun shutdownSinks(timeoutMs: Long = 1_000L): Boolean = true

    /** Returns a subscription that can be closed; no delivery happens in noop. */
    fun addSink(sink: RecordSink): SinkSubscription = SinkSubscription {}

    /** No-op. The real interceptor emits the record to registered sinks. */
    fun injectRecord(record: ContractRecord) = Unit

    override fun close() {
        plugins.removeAll()
    }

    class Builder {
        var maxRequests: Int = 500
        var maxBodySize: Long = 262_144L
        var redactHeaders: Set<String> = HakkaConfig.DEFAULT_REDACTED_HEADERS
        var ignoreHosts: Set<String> = emptySet()
        var ignorePatterns: List<String> = emptyList()
        var sensitiveQueryItems: Set<String> = emptySet()
        var sensitiveBodyFields: Set<String> = emptySet()
        var maxAgeMs: Long? = null
        var enableTiming: Boolean = true
        /**
         * No-op in noop variant. Accepted for API parity with `hakka-network` so that
         * build-flavour switching requires no call-site changes.
         */
        var bridgeUrl: String? = null
        /** No-op in noop variant. API parity with `hakka-network`. */
        var traceEnabled: Boolean = false
        /** No-op in noop variant. API parity with `hakka-network`. */
        var tracePropagateOrigins: List<String> = emptyList()

        private val listeners = mutableListOf<HakkaListener>()
        private val sinks = mutableListOf<RecordSink>()

        fun listener(listener: HakkaListener): Builder = apply { listeners.add(listener) }

        fun sink(sink: RecordSink): Builder = apply { sinks.add(sink) }

        fun build(): HakkaInterceptor {
            return HakkaInterceptor(
                HakkaConfig(
                    maxRequests, maxBodySize, redactHeaders,
                    sensitiveQueryItems, sensitiveBodyFields,
                    ignoreHosts, ignorePatterns, maxAgeMs,
                    bridgeUrl, traceEnabled, tracePropagateOrigins,
                ),
                listeners.toList(),
                sinks.toList(),
            )
        }
    }

    override fun intercept(chain: Interceptor.Chain): Response {
        // Pass through without capturing anything.
        return chain.proceed(chain.request())
    }

    companion object {
        operator fun invoke(block: Builder.() -> Unit = {}): HakkaInterceptor =
            Builder().apply(block).build()
    }
}

File: ios/Sources/NetworkNoop/HakkaInterceptor.swift
import Foundation
import HakkaCommon

/// No-op implementation of `HakkaInterceptor`.
///
/// Drop-in replacement for release builds. All methods are no-ops,
/// `isRunning` is always `false`, and no `URLProtocol` registration
/// or swizzling occurs.
public final class HakkaInterceptor: @unchecked Sendable {
    /// Shared singleton for convenience.
    public static let shared = HakkaInterceptor()

    /// Configuration.
    public let config: HakkaConfig

    /// In-memory log store (no-op).
    public let store: LogStore

    /// Dedicated structured-log ring buffer (no-op build still records entries
    /// locally — logging works even when network capture is disabled).
    public let logStore = HakkaLogStore()

    /// Retention policy (no-op).
    public let retentionPolicy: RetentionPolicy

    /// Delegate for receiving captured requests (never called).
    public weak var delegate: HakkaDelegate?

    /// Closure-based callback (never called).
    public var onRequest: HakkaRequestHandler?

    /// Whether interception is currently active. Always `false`.
    public private(set) var isRunning: Bool = false

    /// Whether capture is paused. Always `false` in the noop implementation.
    public private(set) var isPaused: Bool = false

    public init(config: HakkaConfig = .default) {
        self.config = config
        self.store = LogStore()
        self.retentionPolicy = RetentionPolicy(maxCount: config.maxRequests, maxAge: config.maxAge)
    }

    /// No-op.
    public func start() {}

    /// No-op.
    public func stop() {}

    /// No-op.
    public func pause() {}

    /// No-op.
    public func resume() {}

    /// No-op.
    public func clear() {}

    /// Returns the last `count` captured requests. Always empty.
    public func recentRequests(count: Int = 20) -> [NetworkRequest] { [] }

    /// Returns a snapshot of in-flight requests. Always empty.
    public func inFlightRequests() -> [NetworkRequest] { [] }

    /// Returns a subscription that can be cancelled. No delivery happens in noop.
    public func addSink(_ sink: @escaping RecordSink) -> SinkSubscription {
        SinkSubscription()
    }

    /// No-op.
    @discardableResult
    public func flushSinks(timeout: TimeInterval = 5) -> Bool { true }

    /// Always zero in noop.
    public func droppedSinkRecords() -> Int { 0 }

    /// No-op. The real interceptor emits the record to registered sinks.
    public func inject(_ record: any ContractRecord) {}

    /// Builds an inert health report with no captured records.
    public func healthReport(
        timestamp: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        sessionId: String? = nil,
        tags: [String: String] = [:]
    ) -> HealthReportRecord {
        var healthTags = tags
        healthTags["component.capture.inFlightCount"] = "0"
        healthTags["component.storage.count"] = "0"
        healthTags["component.storage.maxCount"] = String(config.maxRequests)
        healthTags["component.redaction.headerCount"] = String(config.redactHeaders.count)
        healthTags["component.redaction.queryItemCount"] = String(config.sensitiveQueryItems.count)
        healthTags["component.redaction.bodyFieldCount"] = String(config.sensitiveBodyFields.count)

        let redactionConfigured = !config.redactHeaders.isEmpty
            || !config.sensitiveQueryItems.isEmpty
            || !config.sensitiveBodyFields.isEmpty

        return HealthReportGenerator.generate(
            from: [],
            options: HealthReportBuildOptions(
                timestamp: timestamp,
                sessionId: sessionId,
                tags: healthTags,
                componentHealth: [
                    "capture": HealthComponentStatus(status: "disabled"),
                    "storage": HealthComponentStatus(status: "disabled"),
                    "redaction": HealthComponentStatus(status: redactionConfigured ? "configured" : "disabled"),
                    "sink": HealthComponentStatus(status: "disabled", droppedCount: 0),
                ]
            )
        )
    }

    /// No-op. In the real interceptor, this updates mutable config fields
    /// (keeping maxRequests fixed to the ring-buffer size).
    public func updateConfig(_ transform: (HakkaConfig) -> HakkaConfig) {}

    /// Returns an empty metrics summary. The noop interceptor captures nothing.
    public func networkMetricsSummary() -> NetworkMetricsSummary {
        NetworkMetricsSummary(
            totalRequests: 0, completedRequests: 0,
            successCount: 0, errorCount: 0,
            averageResponseTime: 0, successRate: 1.0, errorRate: 0,
            p95LatencyMs: nil, totalDataTransferred: 0
        )
    }

    // MARK: - Structured Logging

    /// Append a structured ``LogEntry`` to ``logStore``. Same API surface as the
    /// real interceptor — logging is never disabled by the network noop swap,
    /// since it doesn't depend on `URLProtocol` interception. Never throws.
    public func log(_ level: LogLevel, _ message: String, category: String? = nil, metadata: [String: String]? = nil) {
        let redacted = metadata.map { config.redactMetadata($0) }
        let entry = LogEntry(level: level, message: message, category: category, metadata: redacted)
        logStore.add(entry)
    }

    /// Convenience: append a debug-level entry.
    public func logDebug(_ message: String, category: String? = nil, metadata: [String: String]? = nil) {
        log(.debug, message, category: category, metadata: metadata)
    }

    /// Convenience: append an info-level entry.
    public func logInfo(_ message: String, category: String? = nil, metadata: [String: String]? = nil) {
        log(.info, message, category: category, metadata: metadata)
    }

    /// Convenience: append a warn-level entry.
    public func logWarn(_ message: String, category: String? = nil, metadata: [String: String]? = nil) {
        log(.warn, message, category: category, metadata: metadata)
    }

    /// Convenience: append an error-level entry.
    public func logError(_ message: String, category: String? = nil, metadata: [String: String]? = nil) {
        log(.error, message, category: category, metadata: metadata)
    }
}

public typealias HakkaDelegate = HakkaCommon.HakkaDelegate
public typealias HakkaRequestHandler = HakkaCommon.HakkaRequestHandler
